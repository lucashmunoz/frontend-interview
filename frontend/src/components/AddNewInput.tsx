import { useState } from "react";
import icons from "../assets/icons";

const { btnAdd } = icons;

interface AddNewInputProps {
  /** Input text value */
  inputValue: string
  /** Callback called on input change event */
  onInputChange: (e: React.FormEvent<HTMLInputElement>) => void
  /** Input placeholder text */
  inputPlaceholder: string
  /** Callback called on add button click */
  onButtonClick: () => void
  /** Button aria label for accessible name */
  buttonAriaLabel: string
}

const AddNewInput = ({
  inputValue,
  onInputChange,
  inputPlaceholder,
  onButtonClick,
  buttonAriaLabel
}: AddNewInputProps) => {
  /*
  * Since the component's border is in the box containing both input and icon button,
  * the input outline has to be removed in favor of changing the color of the box borders
  * on input focus, giving the sense of it being the input outline.
  *
  * If not then the input focus outline will cover only a part of the wrapper box, ending just before the icon button.
  */
  const [isInputOnFocus, setIsInputOnFocus] = useState(false);

  return (
    <div
      className={`border-2 ${isInputOnFocus ? "border-yellow" : "border-black" }  rounded-3xl w-full flex overflow-hidden`}
    >
      <input
        className="w-full py-2 px-4 outline-none"
        placeholder={inputPlaceholder}
        value={inputValue}
        onFocus={() => setIsInputOnFocus(true)}
        onBlur={() => setIsInputOnFocus(false)}
        onChange={onInputChange}
      />
      <button onClick={onButtonClick} aria-label={buttonAriaLabel}>
        <img src={btnAdd} className="h-10 w-10"/>
      </button>
    </div>
  );
};

export default AddNewInput;
