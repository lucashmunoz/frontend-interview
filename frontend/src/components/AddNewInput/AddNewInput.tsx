import { useState } from "react";
import icons from "../../assets/icons";

const { btnAdd } = icons;

interface AddNewInputProps {
  /** Input text value */
  inputValue: string
  /** Callback called on input change event */
  onInputChange: (e: React.FormEvent<HTMLInputElement>) => void
  /** Input placeholder text */
  inputPlaceholder: string
  /** Callback called on form submit */
  onSubmit: () => void
  /** Button aria label for accessible name */
  buttonAriaLabel: string
}

const AddNewInput = ({
  inputValue,
  onInputChange,
  inputPlaceholder,
  onSubmit,
  buttonAriaLabel
}: AddNewInputProps) => {
  /*
  * Since the component's border is in the box containing both input and icon button,
  * the input outline has to be removed in favor of changing the color of the box borders
  * on input focus, giving the sense of it being the input outline.
  *
  * If not then the input focus outline will cover only a part of the wrapper box, ending just before the icon button.
  */
  const [isInputOnFocus, setIsInputOnFocus] = useState(false);

  const handleSubmit = (e: React.SyntheticEvent) => {
    e.preventDefault();

    if(inputValue.length !== 0) {
      onSubmit();
    }
  };

  return (
    <form
      className={`border-2 ${isInputOnFocus ? "border-yellow" : "border-black" }  rounded-3xl w-full flex overflow-hidden`}
      onSubmit={handleSubmit}
    >
      <input
        className="w-full py-2 px-4 outline-none"
        placeholder={inputPlaceholder}
        value={inputValue}
        onFocus={() => setIsInputOnFocus(true)}
        onBlur={() => setIsInputOnFocus(false)}
        onChange={onInputChange}
      />
      <button type="submit" aria-label={buttonAriaLabel}>
        <img src={btnAdd} className="h-10 w-10"/>
      </button>
    </form>
  );
};

export default AddNewInput;
