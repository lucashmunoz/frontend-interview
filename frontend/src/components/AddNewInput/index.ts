import AddNewInput from "./AddNewInput";

export default AddNewInput;
