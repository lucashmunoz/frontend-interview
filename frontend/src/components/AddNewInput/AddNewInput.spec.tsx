import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import AddNewInput from "./AddNewInput";
import { vi } from "vitest";

const mockOnInputChange = vi.fn();
const mockOnSubmit = vi.fn();

describe("AddNewInput", () => {
  afterEach(() => {
    vi.resetAllMocks();
  });

  it("renders the component", () => {
    const { asFragment } = render(
      <AddNewInput
        inputValue="test"
        onInputChange={mockOnInputChange}
        inputPlaceholder="Input Placeholder"
        onSubmit={mockOnSubmit}
        buttonAriaLabel="Button Aria Label"
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should call the onInputChange callback on input change", async () => {
    render(
      <AddNewInput
        inputValue="test"
        onInputChange={mockOnInputChange}
        inputPlaceholder="Input Placeholder"
        onSubmit={mockOnSubmit}
        buttonAriaLabel="Button Aria Label"
      />
    );

    const input = screen.getByRole("textbox");
    await userEvent.type(input, "New input");

    expect(mockOnInputChange).toHaveBeenCalled();
  });

  it("should call the mockOnSubmit callback on button click", async () => {
    render(
      <AddNewInput
        inputValue="test"
        onInputChange={mockOnInputChange}
        inputPlaceholder="Input Placeholder"
        onSubmit={mockOnSubmit}
        buttonAriaLabel="Button Aria Label"
      />
    );

    const button = screen.getByRole("button");
    await userEvent.click(button);

    expect(mockOnSubmit).toHaveBeenCalled();
  });

  it("should not call the mockOnSubmit callback on button click when input is empty", async () => {
    render(
      <AddNewInput
        inputValue=""
        onInputChange={mockOnInputChange}
        inputPlaceholder="Input Placeholder"
        onSubmit={mockOnSubmit}
        buttonAriaLabel="Button Aria Label"
      />
    );

    const button = screen.getByRole("button");
    await userEvent.click(button);

    expect(mockOnSubmit).not.toHaveBeenCalled();
  });
});
