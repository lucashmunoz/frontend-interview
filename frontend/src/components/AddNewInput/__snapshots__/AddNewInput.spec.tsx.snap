// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AddNewInput > renders the component 1`] = `
<DocumentFragment>
  <form
    class="border-2 border-black  rounded-3xl w-full flex overflow-hidden"
  >
    <input
      class="w-full py-2 px-4 outline-none"
      placeholder="Input Placeholder"
      value="test"
    />
    <button
      aria-label="Button Aria Label"
      type="submit"
    >
      <img
        class="h-10 w-10"
        src="/src/assets/btn-add.svg"
      />
    </button>
  </form>
</DocumentFragment>
`;
